test/Microsoft.ML.Tests/ImagesTests.cs

```diff
@@ -184,6 +184,94 @@ public void TestGreyscaleTransformImages()
+        {
+            // Create an image list.
+            var images = new List<ImageDataPoint>(){ new ImageDataPoint(10, 10, Color.Blue), new ImageDataPoint(10, 10, Color.Red) };
+
+            // Convert the list of data points to an IDataView object, which is consumable by ML.NET API.
+            var data = ML.Data.LoadFromEnumerable(images);
+
+            // Convert image to gray scale.
+            var pipeline = ML.Transforms.ConvertToGrayscale("GrayImage", "Image");
+
+            // Fit the model.
+            var model = pipeline.Fit(data);
+
+            // Test path: image files -> IDataView -> Enumerable of Bitmaps.
+            var transformedData = model.Transform(data);
+
+            // Load images in DataView back to Enumerable.
+            var transformedDataPoints = ML.Data.CreateEnumerable<ImageDataPoint>(transformedData, false);
+
+            foreach (var dataPoint in transformedDataPoints)
+            {
+                var image = dataPoint.Image;
+                var grayImage = dataPoint.GrayImage;
+
+                Assert.NotNull(grayImage);
+
+                Assert.Equal(image.Width, grayImage.Width);
+                Assert.Equal(image.Height, grayImage.Height);
+
+                for (int x = 0; x < grayImage.Width; ++x)
+                {
+                    for (int y = 0; y < grayImage.Height; ++y)
+                    {
+                        var pixel = grayImage.GetPixel(x, y);
+                        // greyscale image has same values for R, G and B.
+                        Assert.True(pixel.R == pixel.G && pixel.G == pixel.B);
+                    }
+                }
+            }
+
+            var engine = ML.Model.CreatePredictionEngine<ImageDataPoint, ImageDataPoint>(model);
+            var singleImage = new ImageDataPoint(17, 36, Color.Pink);
+            var transformedSingleImage = engine.Predict(singleImage);
+
+            Assert.Equal(singleImage.Image.Height, transformedSingleImage.GrayImage.Height);
+            Assert.Equal(singleImage.Image.Width, transformedSingleImage.GrayImage.Width);
+
+            for (int x = 0; x < transformedSingleImage.GrayImage.Width; ++x)
+            {
+                for (int y = 0; y < transformedSingleImage.GrayImage.Height; ++y)
+                {
+                    var pixel = transformedSingleImage.GrayImage.GetPixel(x, y);
+                    // greyscale image has same values for R, G and B.
+                    Assert.True(pixel.R == pixel.G && pixel.G == pixel.B);
+                }
+            }
+        }
+
+        private class ImageDataPoint
+        {
+            [ImageType(10, 10)]
+            public Bitmap Image { get; set; }
+
+            [ImageType(10, 10)]
+            public Bitmap GrayImage { get; set; }
+
+            static ImageDataPoint()
+            {
+                DataViewTypeManager.Register(new ImageDataViewType(10, 10), typeof(Bitmap), new ImageTypeAttribute(10, 10));
```

Comment:
Is it really up to the user to register each possible value? It seems like we should be able to create an `ImageDataViewType` instance from any `ImageTypeAttribute`, without the user having to register each size.

Reply:
Do you mean calling register function in a constructor of `ImageTypeAttribute`?

Reply:
If user don't want to specify shapes, they can use `unknown` size image.

---
In reply to: [286694386](https://github.com/dotnet/machinelearning/pull/3263#discussion_r286694386) [](ancestors = 286694386)